inc/abstractrightsdropdown.class.php

```diff
@@ -0,0 +1,252 @@
+      }
+
+      // Add groups if enabled
+      if (self::isTypeEnabled(Group::getType())) {
+         $possible_rights[Group::getType()] = self::getGroups($text);
+      }
+
+      $results = [];
+      foreach ($possible_rights as $itemtype => $ids) {
+         $new_group = [];
+         foreach ($ids as $id => $label) {
+            $new_group[] = [
+               'id' => $id,
+               'text' => $label,
+               "selection_text"=> "$itemtype - $label",
+            ];
+         }
+         $results[] = [
+            "text" => $itemtype::getTypeName(1),
+            "children" => $new_group,
+         ];
+      }
+
+      $ret = [
+         'results' => Toolbox::unclean_cross_side_scripting_deep($results),
+         'count' =>  count($results)
+      ];
+
+      return $ret;
+   }
+
+   /**
+    * Get names for each selected values
+    *
+    * @param array $values Selected values
+    *
+    * @return array
+    */
+   protected static function getValueNames(array $values): array {
+      return array_map(function($value) {
+         $data = explode("-", $value);
+         $itemtype = getItemtypeForForeignKeyField($data[0]);
+         $items_id = $data[1];
+         $item = new $itemtype;
+
+         return $itemtype::getTypeName(1) . " - " . Dropdown::getDropdownName(
+            $item->getTable(),
+            $items_id
+         );
+      }, $values);
+   }
+
+   /**
+    * Get possible values for profiles
+    *
+    * @param string $text Search string
+    *
+    * @return array
+    */
+   protected static function getProfiles(string $text): array {
+      $profile_item = new Profile();
+      $profiles = $profile_item->find([
+         'name' => ["LIKE", "%$text%"]
+      ], [], self::LIMIT);
+      $profiles_items = [];
+      foreach ($profiles as $profile) {
+         $new_key = 'profiles_id-'.$profile['id'];
+         $profiles_items[$new_key] = $profile['name'];
+      }
+
+      return $profiles_items;
+   }
+
+   /**
+    * Get possible values for entities
+    *
+    * @param string $text Search string
+    *
+    * @return array
+    */
+   protected static function getEntities(string $text): array {
+      $entity_item = new Entity;
+      $entities = $entity_item->find(
+         [
+            'name' => ["LIKE", "%$text%"]
+         ] + getEntitiesRestrictCriteria(Entity::getTable()),
+         [],
+         self::LIMIT
+      );
+      $entities_items = [];
+      foreach ($entities as $entity) {
+         $new_key = 'entities_id-'.$entity['id'];
+         $entities_items[$new_key] = $entity['completename'];
+      }
+
+      return $entities_items;
+   }
+
+   /**
+    * Get possible values for users
+    *
+    * @param string $text Search string
+    *
+    * @return array
+    */
+   protected static function getUsers(string $text): array {
+      $users = iterator_to_array(
+         User::getSqlSearchResult(false, "all", -1, 0, [], $text, 0, self::LIMIT)
+      );
```

Comment:
This may consume many memory, depending on the users count. Why not just use a foreach loop on the iterator object ?

Reply:
Its the original code that I pasted here.
Any reason you used iterator_to_array here @orthagh ? Can we change it ?

Reply:
I've removed it